Optimize nsc after reading Hamiltonian from wann90 output

nils-wittemeier opened the pull request:
Before this change, the `nsc` attribute of the Hamtilonian read from wann90 output, was often unnecessarily large. 

The change made in the PR explicitly checks what the maximum range of interactions is after truncation with a specified cutoff. 

Relying on the built-in 'optimize_nsc' is impossible because the Wannier orbitals do not have a specified orbital range.

nils-wittemeier commented:
@zerothi is there a more efficient way of getting a list of all supercell orbitals (i.e. the column indices of all non-zero elements)?

nils-wittemeier commented:
@monx-ne - FYI

zerothi commented:
Perhaps this would be simpler in the `while True` loop, and thus bypass the full thing if the cutoff removes the element? Then we don't need to do the complicated things.

zerothi commented:
also, seems like you have something lying around, it would be great if we could have some tests in sisl which parses a w90 output file :) Any help here would be much appreciated!

zerothi commented:
In fact, much of the code could use a smarter approach. :)

I would advocate a complete re-write, which takes the ideas from `sisl.io.dftb.realdat` which is much simpler, and basically does the same thing. This will result in a much faster processing of the files, and also much faster construction of the full matrix.

nils-wittemeier commented:
> Perhaps this would be simpler in the `while True` loop, and thus bypass the full thing if the cutoff removes the element? Then we don't need to do the complicated things.

Are you just looking for a small test to add to the general sisl tests? - I personally don't have anything, but @monx-ne is using the interface. Maybe we could create a test out of here files.

nils-wittemeier commented:
> In fact, much of the code could use a smarter approach. :)
> 
> I would advocate a complete re-write, which takes the ideas from `sisl.io.dftb.realdat` which is much simpler, and basically does the same thing. This will result in a much faster processing of the files, and also much faster construction of the full matrix.

I can have a look at it in the coming days/next week if you would like me, too.

zerothi commented:
@nils-wittemeier the first thing would just to do the check in the while loop, then the rest can be done later, it is up to you, but any help here would be much appreciated! :)

@monx-ne if you have some small examples in wannier90, it would be useful if you could share them with us, in that way we can add them to the sisl test-suite. :)

zerothi commented:
On another note, w90 seems to have re-implemented a `seedname_tb.dat` output file. This should be a little simpler to parse, so I would suggest you base the implementation on that output file, it should be simpler to parse. :)

nils-wittemeier commented:
That sounds good to me. I'll have a look at the new format.

Would you drop support for the old format or keep both?

nils-wittemeier commented:
Thank you for the hints!

I will get going and if I have any questions I'll let you know.

nils-wittemeier commented:
I reverted the original changes and 
1. Re-implemented the existing reading method similar to `sisl.io.dftb.realdat`
2. Implemented a new reading method for the `_tb.dat` file.

@zerothi please let me know what you think. 

I tested both methods against the old implementation using a silicon example from the [wannier90-tutorials](https://github.com/wannier-developers/wannier-tutorials) and they seem to give the same results.

nils-wittemeier commented:
Rebased on to  `main`

nils-wittemeier commented on src/sisl/io/wannier90/seedname.py:
```diff
@@ -329,43 +330,158 @@
+        #  Lattice vectors [Ang]
+        a = map(float, self.readline())
+        b = map(float, self.readline())
+        c = map(float, self.readline())
```
We could use these vector to check that the `_tb.dat` file is consistent with the geometry, but I'm not sure that's necessary.

nils-wittemeier commented on src/sisl/io/wannier90/seedname.py:
```diff
@@ -329,43 +330,158 @@
+            # Get Hamiltonian matrix elements
+            Hr = Hsc_r[tuple(isc)]
+            Hi = Hsc_i[tuple(isc)]
+            for j in np.arange(no**2):
```
I don't think this matters.

nils-wittemeier commented on src/sisl/io/wannier90/seedname.py:
```diff
@@ -329,43 +330,158 @@
+        Hsc_r = defaultdict(lambda: lil_matrix((geom.no, geom.no), dtype=np.float64))
+        Hsc_i = defaultdict(lambda: lil_matrix((geom.no, geom.no), dtype=np.float64))
+
+        isc = [0, 0, 0]
```
Remove this definition of `isc` in next commit.

zerothi commented:
just let me know when you want me to take over, thanks! :)

nils-wittemeier commented:
I think you can take over already.

nils-wittemeier commented:
One, thing I noticed (and don't quite understand). There are two routines to read the geometry, the first one uses the `_centres.xyz`. The other one seems to try to read the geometry information directly from the input file. This doesn't seem to work correctly in my tests (inconsistent number of orbitals). Should we try to fix that here as well, or in a separate PR?

zerothi commented:
> One, thing I noticed (and don't quite understand). There are two routines to read the geometry, the first one uses the `_centres.xyz`. The other one seems to try to read the geometry information directly from the input file. This doesn't seem to work correctly in my tests (inconsistent number of orbitals). Should we try to fix that here as well, or in a separate PR?

separate pr, thanks!

zerothi commented:
could you share the test you have so we can put it in the CI :)

zerothi commented:
Thanks @nils-wittemeier, it got in!

nils-wittemeier commented:
> could you share the test you have so we can put it in the CI :)

Sure, how do you organize the tests? I saw that there are multiple test scripts and I created one for the reading of the Hamiltonian (explicitly for both file types), but I was unsure where to place the files, that we try reading.

zerothi commented:
Files should be put in this repo: https://github.com/zerothi/sisl-files
then the structure is more or less defined as in sisl. Then you should use the fixture `sisl_files` which will access the correct placement of the files. Then tests can easily be added. :)
Let me know if you have questions, for inspiration, see in `io/siesta/tests/test_fa.py` which is minimal, and shows how to get the `si_pdos_kgrid.FA` file in the `sisl-files` repo.
